### PR TITLE
Add vw and vh units to the custom font size picker

### PR DESCRIPTION
**Pull request by carolinan**
## What?

Add the `vw` and `vh` units to the custom font size picker.
Partial for https://github.com/WordPress/gutenberg/issues/23323

## Why?

The custom font size picker is limited to using px, em, rem. Users are asking for the units to be extended (The units are also more limited than in the custom spacing control, where px, em, rem %, vw and vh are supported)

## How?

The PR updates the units in the font size picker component.

## Testing Instructions
Add a text based block like a paragraph.
Open the block settings and open the Typography panel.
Select "Set custom size". 
Confirm that vw and vh units can be selected.
Select them and confirm that the units are used in the editor and front.

### CODE REVIEW

**Comment by carolinan** [comment]
@ramonjd Hi, I am pinging you early because I think I am missing something. I am trying to add the vw and vh units to the font size picker, and I don't think it can be as easy as this? I can't help thinking that, if it only required this code change, why did we not add it earlier? There must be a rabbit hole somewhere :)

**Comment by ramonjd** [comment]
Thanks for the ping @carolinan 

> I don't think it can be as easy as this? 
> There must be a rabbit hole somewhere :)

How many times have I thought this too 😄 ?

The PR tests well in the editor. I can't think of any technical reason why we couldn't support these units. `vw` already exists as a back up in the [unit control component](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/unit-control/index.js#L17).

Not to say there aren't rabbit holes, yet from a code perspective I can't see any obvious ones.

The only issue that occurs to me right now is that [fluid typography](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/typography.php#L261) doesn't yet support `vw/vh` due to the way the values are calculated. Relative values like these are tricky to get right. 

So a user might be left wondering why some units are converted to clamp values and others not. I suppose one day this could be communicated via the UI somehow, or integrated into any controls we might consider, e.g., 

- https://github.com/WordPress/gutenberg/issues/45502

Pinging @jasmussen for a design perspective.

**Comment by carolinan** [comment]
I suppose we could ask the community if this would be useful without clamp(). I'll ask in the issue.

**Comment by jasmussen** [comment]
The motivation for fluid typography is one of "it just works", and both vh and vw units are fluid in their own way. To that end I'd think this could be as simple as just allowing those units as Carolina suggests. 

There's still a very separate question around how the initial theme.json font size array can be built using the editor, and since theme.json allows a font size preset to be fluid or not, that's something you need to be able to build using global styles. That's mostly a UI design question, but it might eventually overlap with some of aspects of this conversation.

**Comment by mirka** [comment]
Just to first confirm that we're making the intended changes — do we want to expand the unit types that are available in the Typography panel only? Or expand the default unit types for everybody that uses the FontSizePicker component? The PR description suggests that the intent was the former, but the code changes here will do the latter.

Let me know and I can advise on where to make the changes.

**Comment by carolinan** [comment]
🤔 Great question. I think these units are relevant enough to be added to the component...
But open to listen to other opinions.

**Review by mirka — COMMENTED** [review]
I noticed that there is some [`isValueUnitRelative` logic](https://github.com/WordPress/gutenberg/blob/6527034b9e00c4a1297f5fcf5d11e5a10dfbaa02/packages/components/src/font-size-picker/index.tsx#L112) that determines whether to tweak the `max` and `step` values. `vw` and `vh` are both relative values — can we bunch them in with `em` and `rem` for this same logic, or do we want other values for `max` and `step` for these units?

https://github.com/WordPress/gutenberg/blob/6527034b9e00c4a1297f5fcf5d11e5a10dfbaa02/packages/components/src/font-size-picker/index.tsx#L253-L254

**Comment by mirka** [comment]
> I think these units are relevant enough to be added to the component

I wouldn't consider this a "breaking" change so I don't mind either way, but I'm just wondering whether `vw`/`vh` would be kind of irrelevant outside of a "full-page editing" context.

I also wanted to make sure it was clear that we're changing the _defaults_. Any consumer of FontSizePicker could've already passed `vw`/`vh` to the `units` prop if they wanted. I proposed a follow-up because this was unclear from the docs/code #60996.

**Comment by carolinan** [comment]
For the range slider, it makes sense to me to limit the max value to 10 instead of 100, not least because starting at 50 would be very large.
But I am unsure about using the decimals. I can't say I have seen many examples of that. What do you think @iamtakashi ?

What about the native file? I probably should not have touched it @WordPress/native-mobile.

**Comment by carolinan** [comment]
Here is the range slider in the custom font size control in the Typography panel, with the decimals and max value at 10.

https://github.com/WordPress/gutenberg/assets/7422055/f41d7289-ce1f-4a29-a45e-d46e2e7f6700

**Comment by geriux** [comment]
> What about the native file? I probably should not have touched it @WordPress/native-mobile.

I'll take a look 👍

**Comment by geriux** [comment]
All working well on mobile, thanks for the ping! 🚀

**Comment by carolinan** [comment]
Since I have not heard anything, I added vw and vh to `isValueUnitRelative`. It now uses decimals and the range slider has a max value of 10, like em and rem.

**Comment by ramonjd** [comment]
Thanks for the updates. Range slider looks good! 

This is testing as expected for me, and I think a nice addition to the text block font size options. 

Also I can't find any regressions in the way `vw/h` units are parsed when I define them theme.json styles - the values are being reflected in the global styles controls as defaults:

Screenshot 2024-05-09 at 11 01 04 am

I can't see any interference with fluid typography (which guards against units it doesn't understand anyway)

https://github.com/WordPress/gutenberg/assets/6458278/e92d7517-35d0-43f0-9c73-44c66392293d

LGTM, but I'd just cc @WordPress/gutenberg-design for awareness and in case there are any gotchas.

**Comment by jameskoster** [comment]
Seems good to me!

**Review by ramonjd — APPROVED** [review]
Thanks for the PR!

LGTM

Might be worth rebasing to check for changelog conflicts before merging (I had to resolve some locally on this branch)

**Comment by carolinan** [comment]
Which is the correct place to add the change log entry? I understand that I need to move it to _unreleased_, but should it be under enhancements or internal?

File: packages/components/CHANGELOG.md [hunk 1]
```diff
@@ -45,6 +45,7 @@
 
 ### Enhancements
 
+-   `UnforwardedFontSizePicker`: Add `vw` and `vh` units to the font size picker ([#60507]((https://github.com/WordPress/gutenberg/pull/60607)).
```
Comment (t-hamano) [inline packages/components/CHANGELOG.md]:
```suggestion
-   `FontSizePicker`: Add `vw` and `vh` units to the font size picker ([#60507]((https://github.com/WordPress/gutenberg/pull/60607)).
```

This PR "adds" a new unit and is available to users, so I think this section is fine.

**Comment by ramonjd** [comment]
My vote would be under "Enhancements" 👍🏻

**Review by t-hamano — COMMENTED** [review]
I think we will also need to update the JSDoc and README for this component.

https://github.com/WordPress/gutenberg/blob/31d9f7feaf742daff18392a4ea9fec019e7d018c/packages/components/src/font-size-picker/types.ts#L29-L34

https://github.com/WordPress/gutenberg/blob/enhancement/shorcut-selected-blocks-to-group/packages/components/src/font-size-picker/README.md#units-string

**Comment by t-hamano** [comment]
The only concern with this PR is, as already mentioned, changing the default units of the FontSizePicker itself.

And as far as I know, unlike `settings.spacing.units`, the font size units cannot be changed via theme.json. That's because the typography panel doesn't take units into account, and support like `settings.typograpy.units` doesn't exist yet. Therefore, all themes are affected, as well as developers using this component without the `unit` prop.

What do you think about changing the default units of the typography panel in anticipation of adding support for `settings/typograhy.units` in the future?

File: packages/components/CHANGELOG.md [hunk 2]
```diff
@@ -45,6 +45,7 @@
 
 ### Enhancements
 
+-   `FontSizePicker`: Add `vw` and `vh` units to the font size picker ([#60507]((https://github.com/WordPress/gutenberg/pull/60607)).
```
Comment (t-hamano) [inline packages/components/CHANGELOG.md]: Sorry, if you were to ship this PR as is, you would need to move the entry to the "Enhancements" section within the "Unreleased" section.

**Comment by carolinan** [comment]
Without a way to control the units, it is not going to make a difference for either type of extenders where the units are added.

**Comment by t-hamano** [comment]
I'm concerned about backward compatibility and extensibility.

For example, the following changes do not require any changes to the `FontSizePicker` component.

```diff
diff --git a/packages/block-editor/src/components/global-styles/typography-panel.js b/packages/block-editor/src/components/global-styles/typography-panel.js
index 64a7be0443..75e5912d2b 100644
--- a/packages/block-editor/src/components/global-styles/typography-panel.js
+++ b/packages/block-editor/src/components/global-styles/typography-panel.js
@@ -196,6 +196,7 @@ export default function TypographyPanel( {
        const resetFontFamily = () => setFontFamily( undefined );
 
        // Font Size
+       const units = [ 'px', 'em', 'rem', 'vw', 'vh' ];
        const hasFontSizeEnabled = useHasFontSizeControl( settings );
        const disableCustomFontSizes = ! settings?.typography?.customFontSize;
        const mergedFontSizes = getMergedFontSizes( settings );
@@ -385,6 +386,7 @@ export default function TypographyPanel( {
                                                disableCustomFontSizes={ disableCustomFontSizes }
                                                withReset={ false }
                                                withSlider
+                                               units={ units }
                                                size="__unstable-large"
                                        />
                                </ToolsPanelItem>
```

This means that developers using `FontSizePicker` components that do not explicitly specify the `unit` prop will not be affected.

And assuming that font size units are supported in theme.json in the future, all we need to do is make the following changes:

```diff
diff --git a/packages/block-editor/src/components/global-styles/typography-panel.js b/packages/block-editor/src/components/global-styles/typography-panel.js
index 75e5912d2b..38c501d7ee 100644
--- a/packages/block-editor/src/components/global-styles/typography-panel.js
+++ b/packages/block-editor/src/components/global-styles/typography-panel.js
@@ -196,7 +196,8 @@ export default function TypographyPanel( {
        const resetFontFamily = () => setFontFamily( undefined );
 
        // Font Size
-       const units = [ 'px', 'em', 'rem', 'vw', 'vh' ];
+       const [ availableUnits ] = useSettings( 'spacing.units' );
+       const units = availableUnits || [ 'px', 'em', 'rem', 'vw', 'vh' ];
        const hasFontSizeEnabled = useHasFontSizeControl( settings );
        const disableCustomFontSizes = ! settings?.typography?.customFontSize;
        const mergedFontSizes = getMergedFontSizes( settings );
```
What do you think about this approach?

**Comment by carolinan** [comment]
My _opinion_ remains unchanged, that these units are relevant enough to be enabled as the default.

**Comment by richtabor** [comment]
@t-hamano is there any reason why `FontSizePicker` shouldn't have the `vw` and `vh` as default font sizes? If we wanted to use `useSettings( 'spacing.units' )` in the future for font sizes, why not have it there?

**Comment by t-hamano** [comment]
> is there any reason why `FontSizePicker` shouldn't have the `vw` and `vh` as default font sizes?

My only concern is that for developers using the `FontSizePicker` component, and in all themes, the `vw` and `vh` units will be exposed. If many contributors think it's okay for this unit to be exposed as the default, I think it's okay to ship this PR.

**Review by t-hamano — APPROVED** [review]
If we update the CHANGELOG pointed out in [this comment](https://github.com/WordPress/gutenberg/pull/60607/files#r1596281877), we should be able to merge.

**Pull request merged**